event-loop.go

```diff
@@ -0,0 +1,120 @@
+	priorityqueue *priorityqueue.PriorityQueue
+	// called when a task has been received, used to Wakeup the rendering event loop
+	postEmptyEvent func()
+
+	onExpiredTask func(*embedder.FlutterTask) embedder.Result
+
+	// timeout for non-Rendering events that needs to be processed in a polling manner
+	maxWaitDuration time.Duration
+}
+
+func newEventLoop(postEmptyEvent func(), onExpiredTask func(*embedder.FlutterTask) embedder.Result) *EventLoop {
+	pq := priorityqueue.NewPriorityQueue()
+	heap.Init(pq)
+	return &EventLoop{
+		priorityqueue:  pq,
+		postEmptyEvent: postEmptyEvent,
+		onExpiredTask:  onExpiredTask,
+
+		// 25 Millisecond is arbitrary value, not too high (adds too much delay to
+		// platform messages) and not too low (heavy CPU consumption).
+		// This value isn't related to FPS, as rendering events are process in a
+		// waiting manner.
+		maxWaitDuration: time.Duration(25) * time.Millisecond,
+	}
+}
+
+// RunOnCurrentThread returns if the current thread is the thread used
+// by the GLFW event loop.
+func (t *EventLoop) RunOnCurrentThread() bool {
+	// We call runtime.LockOSThread, this ensure the current the
+	// goroutine will always execute in that thread.
+	return true
+}
+
+// PostTask posts a Flutter engine tasks to the event loop for delayed execution.
+func (t *EventLoop) PostTask(task embedder.FlutterTask, targetTimeNanos uint64) {
+	runtime.LockOSThread()
+
+	taskDuration := time.Duration(targetTimeNanos) * time.Nanosecond
+	engineDuration := time.Duration(embedder.FlutterEngineGetCurrentTime())
+
+	t.priorityqueue.Lock()
+	item := &priorityqueue.Item{
+		Value:    task,
+		FireTime: time.Now().Add(taskDuration - engineDuration),
+	}
+	heap.Push(t.priorityqueue, item)
+	t.priorityqueue.Unlock()
+
+	t.postEmptyEvent()
+}
+
+// WaitForEvents waits for an any Rendering or pending Flutter Engine events
+// and returns when either is encountered.
+// Expired engine events are processed
+func (t *EventLoop) WaitForEvents(rendererWaitEvents func(float64)) {
+	now := time.Now()
+
+	expiredTasks := make([]*priorityqueue.Item, 0)
+	var top *priorityqueue.Item
+
+	t.priorityqueue.Lock()
+	for t.priorityqueue.Len() > 0 {
+
+		// Remove the item from the delayed tasks queue.
+		top = heap.Pop(t.priorityqueue).(*priorityqueue.Item)
+
+		// If this task (and all tasks after this) has not yet expired, there is
+		// nothing more to do. Quit iterating.
+		if top.FireTime.After(now) {
+			heap.Push(t.priorityqueue, top) // push the item back into the queue
+			break
+		}
+
+		// Make a record of the expired task. Do NOT service the task here
+		// because we are still holding onto the task queue mutex. We don't want
+		// other threads to block on posting tasks onto this thread till we are
+		// done processing expired tasks.
+		expiredTasks = append(expiredTasks, top)
+
+	}
+	t.priorityqueue.Unlock()
+
+	// Fire expired tasks.
+	for _, item := range expiredTasks {
+		task := item.Value.(embedder.FlutterTask)
```

Comment:
If our priorityqueue always contains item.Value's of type `embedder.FlutterTask`, then we could change the priorityQueue to specifically contain embedder.FlutterTask's, instead of interface{} values.